src/torrent/connection_manager.cc

```diff
@@ -48,11 +48,69 @@
 #include "exceptions.h"
 #include "manager.h"
 
+#ifdef USE_UDNS
+#include "utils/udnsevent.h"
+#endif
+
 namespace torrent {
 
-// Fix TrackerUdp, etc, if this is made async.
-static ConnectionManager::slot_resolver_result_type*
-resolve_host(const char* host, int family, int socktype, ConnectionManager::slot_resolver_result_type slot) {
+/* Encapsulates whether we do genuine async resolution or fall back to sync. */
+struct AsyncResolver {
+
+  ConnectionManager *m_connection_manager;
+#ifdef USE_UDNS
+  UdnsEvent           m_udnsevent;
+#else
+  struct MockResolve {
+    std::string hostname;
+    int family;
+    resolver_callback *callback;
+  };
+  std::list<MockResolve*> m_mock_resolve_queue;
+#endif
+
+  AsyncResolver(ConnectionManager *cm): m_connection_manager(cm) {}
+
+  void *enqueue_resolve(const char *name, int family, resolver_callback *cbck) {
+#ifdef USE_UDNS
+    return m_udnsevent.enqueue_resolve(name, family, cbck);
+#else
+    MockResolve *mock_resolve = new MockResolve {name, family, cbck};
+    m_mock_resolve_queue.push_back(mock_resolve);
+    return mock_resolve;
+#endif
+  }
+
+  void flush_resolves() {
+#ifdef USE_UDNS
+    m_udnsevent.flush_resolves();
+#else
+    // dequeue all callbacks and resolve them synchronously
+    while (!m_mock_resolve_queue.empty()) {
+      MockResolve *mock_resolve = m_mock_resolve_queue.front();
+      m_mock_resolve_queue.pop_front();
+      m_connection_manager->resolver()(mock_resolve->hostname.c_str(), mock_resolve->family, 0, *(mock_resolve->callback));
+      delete mock_resolve;
+    }
+#endif
+  }
+
+  void cancel_resolve(void *query) {
+#ifdef USE_UDNS
+    m_udnsevent.cancel(static_cast<UdnsQuery*>(query));
+#else
+    MockResolve *mock_resolve = static_cast<MockResolve*>(query);
+    auto it = std::find(std::begin(m_mock_resolve_queue), std::end(m_mock_resolve_queue), mock_resolve);
+    if (it != std::end(m_mock_resolve_queue)) {
+      m_mock_resolve_queue.erase(it);
+      delete mock_resolve;
+    }
+#endif
+  }
+};
```

Comment:
This seems a bit too much to include in this file, consider moving it to a source file in the same place as the udns stuff is.